### DISCUSSION TITLE
Tips for compiling MPD in Windows for Windows

### DISCUSSION THREAD

**Discussion by RichardSchorrig** [MusicPlayerDaemon/MPD#1429] [Q&A]
Hi,
I tried to compile MPD as Windows application without success (always with the python build script). First I tried with cygwin64 (it would not go past the compiler tests), then MSYS2 (it would not find the compiler (?)) and finally I installed WSL (Windows Subsystem for Linux) with an Ubuntu 20.04. This was the most promising approach, but in the end it would complain about the meson build file.

Before I take a look at the meson stuff, I wanted to ask if you guys have any tips? Which environment do you use and how did you install it?

Thanks and best regards,
Richard

**Comment by MaxKellermann** [comment]
I've never tried to build MPD on Windows, and actually I don't have Windows.
Building on WSL is building on Linux, not on Windows, because technically, WSL(2) is just a VM with an original Linux. There's no difference between compiling on Linux and WSL2.

**Comment by RichardSchorrig** [comment]
I found my problem: the g++ was missing on the build machine (not the mingw-g++)...

Maybe I'll provide some step by step instructions for dummies like me.

**Comment by RichSiciliano** [comment]
Please help a bigger dummy than you. I went the same route as you, and even tried Ubuntu-16.04 and Ubuntu-18.04 WSL, still can't get past meson build complaints.

**Reply by MaxKellermann** [reply]
No. Nobody can help if you keep the details of your problem a secret.

**Comment by RichSiciliano** [comment]
This is the error I get -

Traceback (most recent call last):                                                                                        File "../../win32/build.py", line 92, in <module>                                                                         from build.libs import *                                                                                              File "/home/richie/MPD/python/build/libs.py", line 6, in <module>                                                         from build.meson import MesonProject                                                                                  File "/home/richie/MPD/python/build/meson.py", line 45                                                                    """)                                                                                                                      ^                                                                                                                 SyntaxError: invalid syntax 

I am just following the users manual. I've been Using MPD on Gnu-Linux for 20 years, but I don't know much hacking other that what I've learned getting things to work over the years (ndiswrapper, alsa, etc.). I only attended school through the 7th grade, and I skipped from 3rd to 5th, so I don't have much of an education. And although I have learned everything I know about computers reading forums, this is my first post, so anything you can do to help would be appreciated, if you can tell me what information you need to from me.

**Reply by MaxKellermann** [reply]
My psychic powers tell me that you're using a Python version older than 6 years, and that's way too old, and that Python version doesn't understand the syntax used by MPD. MPD uses Python 3.6 syntax, and yours is older than that.
And my psychic powers tell me that your GCC version is too old as well. See https://mpd.readthedocs.io/en/stable/user.html#compiling-from-source for details. TLDR: you need at least GCC version 8, and you have only version 7.
Your outdated operation system doesn't cut it. Update.

**Reply by RichSiciliano** [reply]
You could have just said that I need to elevate my mind. I don't know why I insisted on trying to build it in WSL, but I successfully built in a Kubuntu box (after reading the error logs and installing a few more packages, libtool and yasm being the last 2) and downloaded the win64 folder to my windows machine, but when I open mpd.exe a small windows flashes open and nothing else. So I tried it from cmd thinking it would be verbose and I could see what was wrong but it was the same. I appreciate  your assistance, but I guess this problem is out of your scope.

**Reply by MaxKellermann** [reply]
But... what do you expect to happen when you start `mpd.exe`?
Are you aware that MPD is a headless music player? Do you know what that means?

**Reply by RichSiciliano** [reply]
Yes, I've been using it on a headless server for 10 years. I have it hooked up to my Onkyo receiver playing through Advent speakers, sounds great. MPD is the only thing that handles my huge collection well. frankly, it is the most essential program to me.  I control it from my pc or laptop using Cantata, and if I want to play music locally, I have mpd installed on all of my computers and NFS mount my music library from my headless, and just change Cantata to localhost. I rsync my music collection to my webserver and use Subsonic to play music remotely. So I guess what I need to know is how to start and stop mpd server on windows and where the mpd.conf file is. I thought I was building a package that was going to open and install on windows, and then I'd be able to configure everything.

**Reply by MaxKellermann** [reply]
Windows installation instructions are indeed missing in the manual; you can start MPD as a Windows service, but the easiest way to get started now is to just run it from the command line. As parameter, you can pass the `mpd.conf` path.
The MPD website has a precompiled Windows EXE - is there a specific reason why you want to build your own?

**Reply by RichSiciliano** [reply]
I guess it's a bad habit I developed when I started using MythTv on Suse, I don't think there were any packages for it. Maybe windows is not for me. I don't know how to install, configure, or check the status of, a service. Thanks for your help.

**Reply by RichardSchorrig** [reply]
you can try to start it with
mpd.exe mpd.conf --stderr --no-daemon --verbose
given there is a mpd.conf file right next to the mpd.exe

**Reply by RichSiciliano** [reply]
Thanks Richard, that worked. I just configured it with folder that has 2 CD's in it that I mapped as a network drive, and it is playing locally. Now I can add the rest of my music. Thanks for all of your assistance, Rich

**Reply by davidremmons** [reply]
IF I can ask what version of windows you have MPD running successfully on?

**Reply by RichSiciliano** [reply]
I inherited a Dell laptop with Windows 10 and I upgraded it to 11. I had it working on both 10 and 11, but I was not able to add MPD as service, so that it would start at boot. I created a desktop shortcut for the mpd.exe and pinned it to my taskbar, so I could open it with powershell, and then just open new powershell tabs to work in. After a month, I wiped Windows, and replaced it with GNU/Linux (as Richard Stallman rightfully prefers). It's much easier for me to add MPD as a service, and mount my network drive with NFS on GNU/Linux, and at this stage in life, I don't have time to play around with Windows. But Windows is definitely 100 times better than a Mac.
